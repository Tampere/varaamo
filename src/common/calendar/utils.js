import moment from 'moment';
import get from 'lodash/get';

/**
 * getDefaultSelectedTimeRange()
 * Get default selected time range from reservation data.
 * @returns {Object} selected time range
 */
export const getDefaultSelectedTimeRange = (reservation) => {
  return reservation ? {
    start: reservation.begin,
    end: reservation.end,
  } : null;
};


/**
 * isTimeRangeOverMaxPeriod();
 * Reservation cannot be longer than the resources max period if max period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than max_period or not
 */
export const isTimeRangeOverMaxPeriod = (resource, start, end, isStaff = false) => {
  const period = get(resource, 'max_period', null);

  if (!isStaff && period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') > durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * isTimeRangeUnderMinPeriod();
 * Reservation must longer than the resources min period if min period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than min_period or not
 */
export const isTimeRangeUnderMinPeriod = (resource, start, end) => {
  const period = get(resource, 'min_period', null);

  if (period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') < durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * addPeriodToTime()
 * Add period to input time
 * For example: period 00:30:00, time: moment(12:00:00) -> moment(12:30:00)
 *
 * @param {Moment} time
 * @param {string} period
 * @returns {Moment}
 */

const addPeriodToTime = (time, period) => {
  if (period) {
    const periodInMinute = moment.duration(period).as('minutes');
    return time.add(periodInMinute, 'minutes');
  }

  // Return original if no period
  return time;
};

/**
 * getMinPeriodTimeRange
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMinPeriodTimeRange = (resource, start, end) => {
  const minPeriod = get(resource, 'min_period', null);

  if (!minPeriod) {
    return {
      start, end,
    };
  }

  const minPeriodEnd = addPeriodToTime(moment(start), minPeriod);

  if (minPeriodEnd.diff(moment(end)) > 0) {
    // time range smaller than min_period, return min_period
    return {
      start,
      end: minPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};

/**
 * getMaxPeriodEndTime
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMaxPeriodTimeRange = (resource, start, end, isStaff) => {
  const maxPeriod = get(resource, 'max_period', null);

  if (!maxPeriod || isStaff) {
    return {
      start, end,
    };
  }

  const maxPeriodEnd = addPeriodToTime(moment(start), maxPeriod);

  if (moment(end).diff(maxPeriodEnd) > 0) {
    // time range is over max_period not allowed, return max_period
    return {
      start,
      end: maxPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};
