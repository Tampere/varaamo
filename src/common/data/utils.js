import get from 'lodash/get';
import values from 'lodash/values';
import pickBy from 'lodash/pickBy';
import camelCase from 'lodash/camelCase';

export const getLocalizedFieldValue = (field, locale, fallback = false) => {
  const localeValue = get(field, locale, null);

  if (localeValue || !fallback) {
    return localeValue;
  }

  return values(field).find(fallbackValue => !!fallbackValue);
};

/**
 * Trim empty values from data object
 *
 * @param {Object} data
 * @returns {Object} Trimmed empty values object
 */

export const parseData = (data) => {
  return pickBy(data, value => value || value === 0);
};

export const toCamelCase = (obj) => {
  if (!obj) return {};

  const camelCasedObj = {};

  Object.keys(obj).forEach((key) => {
    const camelCasedKey = camelCase(key);
    camelCasedObj[camelCasedKey] = obj[key];
  });

  return camelCasedObj;
};
