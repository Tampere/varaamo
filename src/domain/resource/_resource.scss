@import './card/resourceCard';
@import './card/slider/resourceCardSlider';
@import './map/resourceMap';
@import './reservationCalendar/resourceReservationCalendar';
