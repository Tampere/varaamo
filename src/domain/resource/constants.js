export const resourceProductTypes = Object.freeze({
  RENT: 'rent',
  EXTRA: 'extra',
});

export const resourcePriceTypes = Object.freeze({
  HOURLY: 'hourly',
  DAILY: 'daily',
  WEEKLY: 'weekly',
  FIXED: 'fixed',
});
