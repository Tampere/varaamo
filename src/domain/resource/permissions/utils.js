import difference from 'lodash/difference';
import get from 'lodash/get';

import {
  resourceRoles, resourcePermissionTypes, resourcePermissionsByRole, UI_UNIT_STAFF_ROLES,
} from './constants';

const permissionOptions = Object.values(resourcePermissionTypes);
const allValuesInArray = (subset, superset) => difference(subset, superset).length === 0;

export function hasPermissionForResource(resourceRole, requiredPermissions) {
  // undefined roles don't receive any permissions implicitly
  if (!resourceRole) {
    return false;
  }

  if (!Object.values(resourceRoles).includes(resourceRole)) {
    throw Error(`resourceRole needs to be one of: ${Object.values(resourceRoles).join(', ')}.`);
  }

  if (!Array.isArray(requiredPermissions) && typeof requiredPermissions !== 'string') {
    throw Error('requiredPermissions has to be either a string or an array.');
  }

  if (
    Array.isArray(requiredPermissions)
    && !allValuesInArray(requiredPermissions, permissionOptions)
  ) {
    const unknownPermissions = difference(requiredPermissions, permissionOptions);

    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${unknownPermissions.join(', ')}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  if (typeof requiredPermissions === 'string' && !permissionOptions.includes(requiredPermissions)) {
    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${requiredPermissions}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  const rolePermissions = resourcePermissionsByRole[resourceRole];

  if (Array.isArray(requiredPermissions)) {
    return allValuesInArray(requiredPermissions, rolePermissions);
  }

  return rolePermissions.includes(requiredPermissions);
}

const getUserPermissions = resource => (
  get(resource, 'user_permissions', false)
  || get(resource, 'userPermissions', undefined)
);

// Returns the role with the most permissions or null.
export function getUnitRoleFromResource(resource) {
  const userPermissions = getUserPermissions(resource);

  if (!userPermissions) {
    return null;
  }

  if (userPermissions.is_admin || userPermissions.isAdmin) {
    return resourceRoles.UNIT_ADMINISTRATOR;
  }

  if (userPermissions.is_manager || userPermissions.isManager) {
    return resourceRoles.UNIT_MANAGER;
  }

  if (userPermissions.is_viewer || userPermissions.isViewer) {
    return resourceRoles.UNIT_VIEWER;
  }

  return null;
}

export function getIsUnitStaff(unitRole) {
  return UI_UNIT_STAFF_ROLES.includes(unitRole);
}
