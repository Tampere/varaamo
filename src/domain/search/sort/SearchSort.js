import PropTypes from 'prop-types';
import React from 'react';

import injectT from '../../../../app/i18n/injectT';
import SelectFilter from '../filters/filter/SelectFilter';

const UntranslatedSearchSort = ({
  t,
  locale,
  onChange,
  value,
}) => {
  const sortOptions = [
    { label: t('SearchSort.nameLabel'), value: `resource_name_${locale}` },
    { label: t('SearchSort.typeLabel'), value: `type_name_${locale}` },
    { label: t('SearchSort.premiseLabel'), value: `unit_name_${locale}` },
    { label: t('SearchSort.peopleLabel'), value: 'people_capacity' },
  ];
  const option = sortOptions.find(options => options.value === value);
  const label = !option ? t('SearchSort.premiseLabel') : option.label;

  return (
    <div className="app-SearchSort">
      <div className="sr-only" data-testid="sort-order-announcer" role="status">
        {t('SearchSort.sortingStyle') + label}
      </div>
      <SelectFilter
        id="app-Sort"
        isSearchable={false}
        label={t('SearchSort.label')}
        onChange={item => onChange(item.value)}
        options={sortOptions}
        value={value}
      />
    </div>
  );
};

UntranslatedSearchSort.propTypes = {
  t: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string,
  locale: PropTypes.string,
};

export { UntranslatedSearchSort };

export default injectT(UntranslatedSearchSort);
