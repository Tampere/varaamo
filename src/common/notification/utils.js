import { NotificationManager } from 'react-notifications';

import { NOTIFICATION_TYPE } from './constants';
/**
 * createNotification()
 * Creating notification popups
 *
 * @param {string} type
 * @param {string} [message='']
 * @param {string} [title='']
 * @param {number} [fadeTime=2000]
 * @param {function} callback
 * @returns {void}
 */
export const createNotification = (type, message = '', title = '', fadeTime = 2000, callback) => {
  switch (type) {
    case NOTIFICATION_TYPE.INFO:
      NotificationManager.info(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.SUCCESS:
      NotificationManager.success(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.WARNING:
      NotificationManager.warning(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.ERROR:
      NotificationManager.error(message, title, fadeTime, callback);
      break;
    default:
      NotificationManager.info(message, title, fadeTime, callback);
  }
};
