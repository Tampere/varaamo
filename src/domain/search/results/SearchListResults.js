import React from 'react';
import PropTypes from 'prop-types';
import Loader from 'react-loader';
import get from 'lodash/get';
import find from 'lodash/find';
import omit from 'lodash/omit';
import { withRouter } from 'react-router-dom';
import Row from 'react-bootstrap/lib/Row';
import Col from 'react-bootstrap/lib/Col';
import moment from 'moment';

import constants from '../../../../app/constants/AppConstants';
import SearchSort from '../sort/SearchSort';
import Pagination from '../../../common/pagination/Pagination';
import ResourceCard from '../../resource/card/ResourceCard';
import * as searchUtils from '../utils';

class SearchListResults extends React.Component {
  static propTypes = {
    resources: PropTypes.array.isRequired,
    units: PropTypes.array.isRequired,
    isLoading: PropTypes.bool,
    totalCount: PropTypes.number,
    history: PropTypes.object.isRequired,
    location: PropTypes.object.isRequired,
    onFavoriteClick: PropTypes.func.isRequired,
    onFiltersChange: PropTypes.func.isRequired,
  };

  onSortChange = (sort) => {
    const { history, location } = this.props;
    const filters = searchUtils.getFiltersFromUrl(location);

    const newFilters = {
      ...omit(filters, 'orderBy'),
    };

    if (sort) {
      newFilters.orderBy = sort;
    }

    history.push({
      search: searchUtils.getSearchFromFilters(newFilters),
    });
  };

  onFilterClick = (filterName, filterValue) => {
    const { onFiltersChange } = this.props;

    onFiltersChange({ [filterName]: filterValue });
  };

  render() {
    const {
      isLoading,
      resources,
      units,
      location,
      history,
      totalCount,
      onFavoriteClick,
    } = this.props;

    const filters = searchUtils.getFiltersFromUrl(location);

    return (
      <div className="app-SearchListResults">
        <div className="app-SearchListResults__sort">
          <Row>
            <Col md={4} mdOffset={8} sm={6}>
              <SearchSort
                onChange={sort => this.onSortChange(sort)}
                value={get(filters, 'orderBy', '')}
              />
            </Col>
          </Row>
        </div>
        <Loader loaded={!isLoading}>
          <div className="app-SearchListResults__results">
            {resources && resources.map((resource) => {
              const unit = find(units, item => item.id === resource.unit);

              if (unit) {
                return (
                  <ResourceCard
                    date={get(filters, 'date', moment().format(constants.DATE_FORMAT))}
                    key={`resourceCard-${resource.id}`}
                    onFavoriteClick={onFavoriteClick}
                    onFilterClick={this.onFilterClick}
                    resource={resource}
                    unit={unit}
                  />
                );
              }

              return null;
            })}
          </div>
        </Loader>
        <Pagination
          onChange={newPage => history.push({
            search: searchUtils.getSearchFromFilters({ ...filters, page: newPage }),
          })}
          page={filters && filters.page ? Number(filters.page) : 1}
          pages={Math.ceil(totalCount / constants.SEARCH_PAGE_SIZE)}
        />
      </div>
    );
  }
}

const UnwrappedSearchListResults = SearchListResults;

export { UnwrappedSearchListResults };

export default withRouter(SearchListResults);
