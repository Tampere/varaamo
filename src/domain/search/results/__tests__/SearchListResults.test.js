import React from 'react';
import toJSON from 'enzyme-to-json';
import { shallow } from 'enzyme';

import { UnwrappedSearchListResults } from '../SearchListResults';
import resource from '../../../../common/data/fixtures/resource';
import unit from '../../../../common/data/fixtures/unit';
import { globalDateMock } from '../../../../../app/utils/testUtils';
import SearchSort from '../../sort/SearchSort';
import Pagination from '../../../../common/pagination/Pagination';

const findSearchSort = wrapper => wrapper.find(SearchSort);
const findPagination = wrapper => wrapper.find(Pagination);

describe('SearchListResults', () => {
  globalDateMock();

  const defaultProps = {
    history: {},
    location: {},
    match: {},
    onFavoriteClick: () => null,
    onFiltersChange: () => null,
    resources: [resource.build()],
    units: [unit.build()],
  };
  const getWrapper = props => shallow(
    <UnwrappedSearchListResults {...defaultProps} {...props} />,
  );
  test('renders correctly', () => {
    const wrapper = getWrapper();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test('correctly passes orderBy to SearchSort', () => {
    const orderBy = 'value';
    const location = { search: `?orderBy=${orderBy}` };
    const wrapper = getWrapper({ location });

    expect(findSearchSort(wrapper).prop('value')).toEqual(orderBy);
  });

  test('correctly passes pages to Pagination', () => {
    const wrapper = getWrapper({ totalCount: 67 });

    expect(findPagination(wrapper).prop('pages')).toEqual(3);
  });
});
