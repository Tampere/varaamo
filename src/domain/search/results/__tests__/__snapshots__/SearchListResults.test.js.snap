// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchListResults renders correctly 1`] = `
<div
  className="app-SearchListResults"
>
  <div
    className="app-SearchListResults__sort"
  >
    <Row
      bsClass="row"
      componentClass="div"
    >
      <Col
        bsClass="col"
        componentClass="div"
        md={4}
        mdOffset={8}
        sm={6}
      >
        <InjectT(UntranslatedSearchSort)
          onChange={[Function]}
          value=""
        />
      </Col>
    </Row>
  </div>
  <Component
    component="div"
    loaded={true}
    loadedClassName="loadedContent"
    parentClassName="loader"
  >
    <div
      className="app-SearchListResults__results"
    >
      <InjectIntl(Connect(InjectT(ResourceCard)))
        date="2017-12-10"
        key="resourceCard-1"
        onFavoriteClick={[Function]}
        onFilterClick={[Function]}
        resource={
          Object {
            "available_hours": Array [],
            "equipment": Array [],
            "id": 1,
            "is_favorite": false,
            "name": Object {
              "en": "Resource (en) - 1",
              "fi": "Resource (fi) - 1",
              "sv": "Resource (sv) - 1",
            },
            "need_manual_confirmation": false,
            "opening_hours": Array [],
            "products": Array [],
            "public": true,
            "required_reservation_extra_fields": Array [],
            "reservable": true,
            "reservable_after": null,
            "slot_size": "00:30:00",
            "supported_reservation_extra_fields": Array [],
            "unit": 1,
            "user_permissions": Object {
              "canMakeReservations": true,
              "isAdmin": false,
            },
          }
        }
        unit={
          Object {
            "id": 1,
            "location": Object {
              "coordinates": Array [
                25.01571,
                60.249935,
              ],
              "type": "Point",
            },
            "name": Object {
              "en": "Unit (en) - 1",
              "fi": "Unit (fi) - 1",
              "sv": "Unit (sv) - 1",
            },
            "opening_hours_today": Object {},
          }
        }
      />
    </div>
  </Component>
  <InjectT(Pagination)
    onChange={[Function]}
    page={1}
    pages={NaN}
  />
</div>
`;
